MergeSamFiles accept SO:UNKNOWN

a-filipp0v opened the pull request:
### Description

Request is associated with Picard's issue https://github.com/broadinstitute/picard/issues/796

As said in that issue MergeSamFiles can't accept SAM-file with SO:UNKNOWN. That behavior is expected, because tag's value is in improper case(in accordance with SAM-format specification, SO value must be in lowercase). We've found that wrong tag case also provides an exception.

The idea of implementation is that htsjdk must check case of header's tags and values. If tag/value itself is valid, but in improper case(e.g. lowercase instead of uppercase) - it should be converted to proper case before sending the result to output(without changing the source file). Also, there must be a message if that conversion was made.

### Checklist

- [ ] Code compiles correctly
- [ ] New tests covering changes and new functionality
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary
- [ ] Is not backward compatible (breaks binary or source compatibility)

yfarjoun commented:
Personally I'm opposed to this change. 

I do not see why htsjdk needs to be made more complicated in order to accept non-SAM-spec compliant input. Its hard enough as it is even with SAM-compliant input and this will make the code more complicated. What if someone has a typo in the header and they spelled the sort order "unkwnon"? Or perhaps they use a different "locale" and the used "SO: desconocido" ? should htsjdk also accept that? 

That off my chest, I'm not a maintainer...so I'd love to hear from them before even looking at the details of the implementation: @lbergelson @tfenne  @jacarey

yfarjoun commented:
I think that the correct solution is to fix the offending software so that it produces compliant BAM files and to write a simple script that fixes "broken" headers.

yfarjoun commented:
Alternatively a better  (in my view) solution would be an ability to ignore (when validation=silent or leniant) an unknown sort order and use "unknown" when that happens (instead of lowercasing)

magicDGS commented:
I also prefer that HTSJDK is SAM-compliant and do not allow non-supported `SO`...

jacarey commented:
I agree with @yfarjoun

yfarjoun commented:
@a-filipp0v could you modify the code as follows:
1. remove the lower-casing and the logs related to that
2. if SortOrder.valueOf(unmodified) throws an error, AND VALIDATION_STRINGENCY!=STRICT, use SortOrder.UNKNOWN and either emit a warning or not depending on VALIDATION_STRINGENCY.

I think that this will both solve the issue at hand and not introduce un-needed complexity.

As a bonus, I think that your tests should still be useful!

a-filipp0v commented:
Thank you for your comments. Hope we've understood you correctly in case of non-conforming SO tag values.
Our implementation has been reworked according to your requests. Also, we've reworked previous tests in SamFileHeaderTest, added few more and removed one test case from MergeSamFilesTest because it's redundant after our implementation.

yfarjoun commented on src/main/java/htsjdk/samtools/SAMFileHeader.java:
```diff
@@ -324,12 +324,18 @@ public void setAttribute(final String key, final Object value) {
+                this.sortOrder = null;
+                tempVal = SortOrder.valueOf(value).toString();
+            } catch (IllegalArgumentException e) {
+                tempVal = "unknown";
```
use `SortOrder.unknown.toString()` here.

yfarjoun commented on src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java:
```diff
@@ -323,11 +326,30 @@ private void reportErrorParsingLine(String reason, final SAMValidationError.Type
         }
 
+        private void validateSortOrderValue(String[] value) {
+            if ("SO".equals(value[0])) {
```
use SAMFileHeader.SORT_ORDER_TAG

yfarjoun commented on src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java:
```diff
@@ -323,11 +326,30 @@ private void reportErrorParsingLine(String reason, final SAMValidationError.Type
+                        log.warn("Found non conforming header SO tag: "
+                                 + value[1] + ". Treating as 'unknown'.");
+                    }
+                    value[1] = "unknown";
```
use SortOrder.unknown.toString()

yfarjoun commented on src/main/java/htsjdk/samtools/SAMFileHeader.java:
```diff
@@ -324,12 +324,18 @@ public void setAttribute(final String key, final Object value) {
         if (key.equals(SORT_ORDER_TAG)) {
-            this.sortOrder = null;
+            try {
+                this.sortOrder = null;
```
Why move this line into the try clause?

yfarjoun commented on src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java:
```diff
@@ -69,6 +70,8 @@
 
     public static final String COMMENT_PREFIX = HEADER_LINE_START + HeaderRecordType.CO.name() + FIELD_SEPARATOR;
+    private static final Log log = Log.getInstance(SAMTextHeaderCodec.class);
+
```
remove extra newline

yfarjoun commented on src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java:
```diff
@@ -233,8 +236,8 @@ private void parseHDLine(final ParsedHeaderLine parsedHeaderLine) {
         } catch (IllegalArgumentException e) {
-            reportErrorParsingLine(HEADER_LINE_START + parsedHeaderLine.getHeaderRecordType() +
-                            " line has non-conforming SO tag value: "+ soString + ".",
+                reportErrorParsingLine(HEADER_LINE_START + parsedHeaderLine.getHeaderRecordType() +
```
revert extra tab

yfarjoun commented on src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java:
```diff
@@ -323,11 +326,30 @@ private void reportErrorParsingLine(String reason, final SAMValidationError.Type
                             SAMValidationError.Type.HEADER_TAG_MULTIPLY_DEFINED, null);
                     continue;
                 }
+                validateSortOrderValue(keyAndValue);
```
why is this needed here? could you put it instead in SamFileValidator? perhaps in the ValidateHeader method?

note that the validation methods create a tally of validation failures... and this should probably do the same.

yfarjoun commented on src/test/java/htsjdk/samtools/ValidateSamFileTest.java:
```diff
@@ -552,12 +552,6 @@ public void duplicateReadsOutOfOrder() throws Exception {
                         "@RG\tID:0\tSM:Hi,Mom!\n" +
                         "E\t147\tchr1\t15\t255\t10M\t=\t2\t-30\tCAACAGAAGC\t)'.*.+2,))\tU2:Z:CAA";
 
-        final String SOTagCorrectlyProcessTestData =
```
why is this removed?

IanaKabakova replied:
thank you for your reply. We've removed this test case since as @a-filipp0v said, it's redundant after our implementation.

yfarjoun replied:
not sure I understand...if it's just "Redundant" but not exactly equals, I'd ask that it be left in.

IanaKabakova replied:
This case does not pass now, since we no longer treat SO:NOTKNOWN as error as test expects.

yfarjoun replied:
got it

a-filipp0v commented:
Hello, @yfarjoun , thanks for your reply.

According to your previous comments, we've tried to move an appropriate check to SamFileValidator, but found that it still might make sense to leave our checks in SAMTextHeaderCodec and SAMFileHeader.

First, the check in SamFileValidator will be called only when the ValidateSamFile is used, which does not prevent an error when using any tools separately (for example, MergeSamFile as in this issue).

Secondly, SamFileValidator does not contain separate data about Validation Stringency, because this parameter is parsed from Picard during start.

We suggest leaving checks in SAMTextHeaderCodec and SAMFileHeader AND adding a similar one in SamFileValidator. What do you think about it?

lbergelson commented:
@yfarjoun Since you did the initial review, could you take a look at this again when you get a chance?

yfarjoun commented:
@a-filipp0v thanks for the explanation. I've removed my comment relating to the validation.

yfarjoun commented:
@a-filipp0v sorry for the delay. please respond to the review comments.

yfarjoun commented:
@a-filipp0v just so we know, are you planning to get back to this PR?

IanaKabakova commented:
@yfarjoun sorry for the delay and thank you for your reply. I made some fixes according to your notes. Could you take a look at PR again?

yfarjoun commented:
Thanks for the changes. Please leave tests as is, unless they no longer pass.

yfarjoun commented:
@jacarey @tfenne @lbergelson any opinions about this small API change?

pshapiro4broad commented on src/main/java/htsjdk/samtools/SAMFileHeader.java:
```diff
@@ -324,12 +324,18 @@ public void setAttribute(final String key, final Object value) {
         if (key.equals(SORT_ORDER_TAG)) {
             this.sortOrder = null;
+            try {
+                tempVal = SortOrder.valueOf(value).toString();
```
In `setSortOrder()` the code uses `.name()` instead of `.toString()` which I think is clearer; either way the code should be consistent.

Also, it seems odd that `setSortOrder()` sets `sortOrder` to non-null, if the argument is a valid `SortOrder`, but this method sets it to `null`. Maybe this method should just call `setSortOrder()` when `key` is `SORT_ORDER_TAG`?

yfarjoun replied:
@pshapiro4broad that's how the old implementation also does it...it let's the super-class deal with it...take a look at htsjdk.samtools.SAMFileHeader#getSortOrder, I think that might clarify things up (and note that this.sortOrder is private)

pshapiro4broad commented on src/main/java/htsjdk/samtools/SAMTextHeaderCodec.java:
```diff
@@ -323,11 +326,30 @@ private void reportErrorParsingLine(String reason, final SAMValidationError.Type
             lineValid = true;
         }
 
+        private void validateSortOrderValue(String[] value) {
```
I think it would be slightly clearer if the caller extracted `value[0]` and `value[1]` so they could be declared as `String key` and `String value` here.

yfarjoun commented:
No response from the maintainers for a while now, I suspect that no-one objects....if I'm wrong, please revert and accept my apologies.
